XLS-0066d-lending-protocol/README.md

```diff
@@ -0,0 +1,1428 @@
+
+$$
+valueChange = interestLate - interestPeriodic
+$$
+
+Note that `valueChange >= 0`.
+
+###### 3.2.5.1.3 Loan Overpayment
+
+- Let $\mathcal{P}$ and $\mathcal{p}$ represent the total and outstanding Loan principal.
+- Let $\mathcal{I}$ and $\mathcal{i}$ represent the total and outstanding Loan interest computed from $\mathcal{P}$ and $\mathcal{p}$ respectively.
+
+$$
+excess = min(\mathcal{p}, paymentAmountMade - minimumPaymentAmount)
+$$
+
+$$
+interestPortion = excess \times overpaymentInterestRate
+$$
+
+$$
+feePortion = excess \times overpaymentFee
+$$
+
+$$
+principalPortion = excess - interestPortion - feePortion
+$$
+
+$$
+\mathcal{p'} = \mathcal{p} - principalPortion
+$$
+
+Let $\mathcal{i}$ denote the outstanding interest computed from $\mathcal{p}$. Simillarly, let $\mathcal{i'}$ denote the outstanding interest computed from $\mathcal{p'}$. We compute the loan interest change as follows:
+
+$$
+valueChange =  \mathcal{i} - \mathcal{i'}
+$$
+
+###### 3.2.5.1.4 Early Full Repayment
+
+A Borrower can close a Loan early by submitting the total amount needed to do so. This amount is the sum of the remaining balance, any accrued interest, a prepayment penalty, and a prepayment fee.
+
+$$
+totalDue = principalOutstanding + accruedInterest + prepaymentPenalty + ClosePaymentFee
+$$
+
+Accrued interest up to the point of early closure is calculated as follows:
+
+$$
+accruedInterest = principalOutstanding  \times periodicRate \times \frac{secondsSinceLastPayment}{paymentInterval}
+$$
+
+Finally, the Lender may charge a prepayment penalty for paying a loan early, which is calculated as follows:
+
+$$
+prepaymentPenalty = principalOutstanding \times closeInterestRate
+$$
+
+An early payment pays less interest than calculated when increasing the Vault value in the `LoanSet` transaction. Therefore, the Vault value (captured by `Vault.AssetTotal`) must be recalculated after an early payment.
+
+Assume a function `CurrentValue()` that returns `principalOutstanding` and `interestOutstanding` of the Loan. Furthermore, assume a function `ClosePayment()` that implements the Full Payment calculation. The function returns the total full payment due split into `principal` and `interest`.
+
+The value change for an early full repayment is calculated as follows:
+
+$$
+valueChange = (prepaymentPenalty) - (interestOutstanding - accruedInterest)
+$$
+
+Note that `valueChange <= 0` as an early repayment reduces the total value of the Loan.
```

Comment:
Are there any circumstances where `prepaymentPenalty > (interestOutstanding - accruedInterest)` such that `valueChange > 0`?

Reply:
> In theory, yes, it is possible to have `valueChange > 0`.
> 
> Remember: `prepaymentPenalty = principalOutstanding \times closeInterestRate`
> 
> `valueChage > 0` suggests that the borrower would pay MORE to close the Loan early, than repay the loan in scheduled payments.

Yeah, those would be pretty cruddy terms to accept on a loan. So do we make `valueChange <= 0` a hard requirement to stop loan sharking? 😁